### ISSUE TITLE
Opvolging van SLACK moet de API de flow ZAAK, EIO zonder inhoud & ZIO ondersteunen of niet?

### ISSUE THREAD

**Issue by shazada** [VNG-Realisatie/gemma-zaken#2282]
Zie slack thread:
https://samenorganiseren.slack.com/archives/CT6UH711Q/p1688464270356929

Oorspronkelijke vraag:
Weet iemand of je bij een gelockte DRC document wel een ZIO (Zaak-Informatieobject kunt aanmeken)?
1.8.2 interne DRC werkt dit wel, Gemma DRC 1.3.0 werkt dit niet, onze Alfresco/Tezza-DRC gaat dit ook stuk.
Foutmelding staat hier https://github.com/open-zaak/open-zaak/issues/1386#issuecomment-1607355553, maar de API is niet duidelijk in de flow.
Dus functioneel is de flow:

1. EIO aanmaken in DRC, zonder inhoud dus doc is gelockt
2. ZIO aanmaken in ZRC, dus zaak-informatieobject
3. EIO bestandsdelen uploaden en daarna is de doc niet meer gelockt

Bevinding is per combi DRC en Combi ZRC dus anders.

Bevinding in Openzaak github: https://github.com/open-zaak/open-zaak/issues/1386#issuecomment-1607355553

**Comment by johannesbattjes** [comment]
Toelichting vanuit Rx.Mission. In de basis is dit de flow die Rx.Mission hanteert:

- Registreer enkelvoudiginformatieobject (alleen meta-data dus)
- Koppel aan zaak
- Voeg bestandsinhoud toe

De reden voor deze volgorde is eenvoudig: stap 3) kan enige tijd duren en kan potentieel fout gaan. Het gaat niet altijd om een simpele upload van een documentje, het kan gaan om een complexe documentgeneratie waar meerdere databronnen bij betrokken zijn of om de upload van een zeer groot bestand. Dus het is nodig om gedurende stap 3) de voortgang aan de eindgebruiker te kunnen tonen en mocht er iets mis gaan moet dat ook getoond worden. Dat kan alleen als de relatie met de zaak al gelegd is, anders weet het systeem helemaal niet dat dit document bestaat bij die zaak.

Bovendien kan het zijn dat er bij automatisch gegenereerde documenten data in stap 3) nodig is die pas bekend is na stap 1). Het documentnummer wil je bijvoorbeeld kunnen weergeven in een gegenereerde brief, maar het documentnummer wordt automatisch bepaald door de DRC bij het registreren van het document in stap 1).

**Comment by michielverhoef** [comment]
Zonder alvast een uitspraak te willen doen lijkt deze flow heel aardig op wat vroeger in Zaak- Documentservices gebruikt werd:

- genereerDocumentidentificatie
- voegZaakdocumentToe (met inhoud) 

of

- genereerDocumentidentificatie
- maakZaakdocument (zonder inhoud, bijvoorbeeld in geval van documentgeneratie)
- updateZaakdocument (inhoud toevoegen)

Weliswaar is het niet 1 op 1 identiek maar het idee is vergelijkbaar. Er wordt eerst een documentnummer gegenereerd en gereserveerd en daarna wordt het document aan de zaak toegevoegd. 

Wat ik niet goed zie nog is: Bij het aanmaken van een enkelvoudig informatieobject (EIO) met behulp van bestandsdelen moet de grootte van het EIO meegegeven worden. Dit zodat de Documenten API kan berekenen hoeveel bestandsdelen er moeten komen.  Wanneer een document nog gegenereerd moet worden kan het gebeuren dat die grootte nog niet bekend is. Wordt dan eigenlijk alleen de metadata opgeslagen, zonder (grootte van de ) inhoud? Dan wordt eigenlijk het patroon van de tweede optie (maakZaakdocument + updateZaakdocument) gevolgd.

**Comment by MNIJ** [comment]
@michielverhoef, inderdaad wordt in stap 1 alleen de metadata geregistreerdt, dus zonder bestandsinhoud en zonder bestandsomvang. Dat is ook gewoon toegstaan in de standaard, de velden `inhoud` en `bestandsomvang` zijn nullable.

Het toevoegen van de bestandsinhoud gebeurt dus met een PUT of PATCH. Of dat in bestandsdelen gaat of met een BASE64 vulling van het `inhoud` veld ligt aan de bestandsgrootte en de use-case.

Overigens is het voor stap 1 en stap 2 niet nodig (of zelfs niet mogelijk) om het EIO te locken. Dus ik vraag me af of de stelling van @shazada klopt dat het EIO in stap 1 gelockt wordt, volgens mij gebeurt dat pas bij stap 3.

**Comment by shazada** [comment]
@shazada Functioneel weet ik niet hoe het zou moeten werken, maar de DRC api geeft dit weer:
als je in stap 1 geen inhoud meegeeft:

`    "locked": true,
    "bestandsdelen": [
        {
            "url": "http://drc.local:8000/api/v1/bestandsdelen/3782580e-db9b-41cd-92f2-697dd5a61c88",
            "volgnummer": 1,
            "omvang": 47185920,
            "voltooid": false
        }
    ],`
    
    Dus de DRC blijft wachten totdat het document inhoud heeft en volledig is.

**Comment by MNIJ** [comment]
Ja, excuus. In de flow van een directe upload van een groot bestand wordt wel een bestandsomvang in stap 1 meegestuurd en wordt deze wel direct in stap 1 gelockt.

**Comment by johannesbattjes** [comment]
Ik vraag me af of in dit issue het juiste probleem wordt onderzocht.
We hebben (vanuit Postman) een aantal testscenario's doorlopen met het aanmaken of bijwerken van documenten op basis van bestandsdelen. In alle drie de scenario's gaat het op Open Zaak probleemloos.
Bij de referentie-implementatie gaat (alleen) het aanmaken ZIO mis. Tot zover lijken onze bevindigen op die van Tahir. 
Maar het aanmaken van ZIO gaat ook mis inn de referentie implementatie als het document al met inhoud bestaat en niet gelocked is. Zie hieronder een samenvatting. Bij de rode streepjes gaat het dus mis in de referentie implementatie (maar niet Open Zaak).

Voorlopige conclusie: er is iets mis met het aanmaken van ZIO in de rerenetieimplementatie.
Foutmelding:
{
    "type": "https://zaken-api.vng.cloud/ref/fouten/ValidationError/",
    "code": "invalid",
    "title": "Invalid input.",
    "status": 400,
    "detail": "",
    "instance": "urn:uuid:917657b5-2e89-4799-8097-93531b9d1e07",
    "invalidParams": [
        {
            "name": "nonFieldErrors",
            "code": "invalid",
            "reason": "Could not create remote relation"
        }
    ]
}


![image](https://github.com/VNG-Realisatie/gemma-zaken/assets/45750838/99d7df86-ad05-4086-8de4-68c5677859a5)

**Comment by shazada** [comment]
Na de analyse van Alex van Maykin zijn we erachter gekomen los van het functionele dat er en technische verschil zit in de bestanddelen en het locked attribuut in de DRC versies, zie ticket: https://github.com/VNG-Realisatie/gemma-zaken/issues/2293

@michielverhoef kun jij nog functioneel antwoorden of de flow wel/niet wordt gesupport en wil antwoorden waarom de locked attribuut in versie 1.1 is geïntroduceerd en in 1.2 weg is gehaald en in 1.3 weg is gebleven? Ik heb je net ook geprobeerd te bellen, zou je me anders willen terugbellen hierover? Of misschien een meeting inplannen met de betrokkenen?

**Comment by michielverhoef** [comment]
@shazada in mijn out-of-office bericht stond dat ik vakantie had en niet aanwezig was.  ;-)

@johannesbattjes  schreef:
> Voorlopige conclusie: er is iets mis met het aanmaken van ZIO in de refenetie-implementatie.

Dit klinkt als een bug die opgelost moet worden.

@shazada schreef:
> kun jij nog functioneel antwoorden of de flow wel/niet wordt gesupport en wil antwoorden waarom de locked attribuut in versie 1.1 is geïntroduceerd en in 1.2 weg is gehaald en in 1.3 weg is gebleven? 


Dit moet ik uitzoeken. Vermoedelijk is dit geen bewuste keuze geweest maar een fout in de implementatie. Het is wel de eerste keer dat dit gemeld wordt, dank hiervoor.

**Comment by shazada** [comment]
@michielverhoef Je hebt gelijk dit kan gewoon in dit ticket. zou je hier aub nog op willen reageren?
We hebben 1 gemeente die nu alles still heeft gelegd omdat dit onduidelijk is en nog niet is gefixt. 

Uit de flow lijkt dat de locked attribuut dus in 1.4.0 ook weg is lijkt me. Kunnen we hiermee accepteren dat de dubbele locked attribuut dus in de Alpha versie die 1.1 werkt verkeerd is geïntroduceerd?

Functioneel is het ook onzinnig lijkt het mij, want je hebt
- EIO heeft een locked attribuut
- binnen bestandsdelen heb je een voltooid attribuut met true/false
- om hierbinnen bovenop ook een locket attribuut te hebben heeft geen toegevoegde waarde
